### PR TITLE
[Relay][Frontend][Onnx] GRU Layer Support

### PR DESCRIPTION
**Pull request by jwfromm**
This PR adds GRU parsing to the onnx frontend. Currently this is done by unrolling the recurrence in a similar way to how we handle LSTMs. Since there's quite a bit of shared code, I've generalized the LSTM converter to an RNN converter and made LSTM and GRU subclasses. For testing, I again replaced the LSTM test function with a more general test_rnn function that supports both LSTM and GRU and should be easily expandable to other RNN functions should we ever want to add any.

### CODE REVIEW

**Comment by jwfromm** [comment]
@masahi, @areusch can you guys take a look at this PR?

**Comment by anijain2305** [comment]
This might be somewhat relevant - https://discuss.tvm.ai/t/onnx-lstm-op-conversion/7238
Just pointing it out here. I think the GRU might also have the same problems that I encounter with LSTM.

**Comment by masahi** [comment]
@siju-samuel please help review this PR.

**Comment by jwfromm** [comment]
I've converted this to a draft as I'll try to incorporate the points made by @anijain2305 in the linked discuss post.

File: python/tvm/relay/frontend/onnx.py [hunk 1]
```diff
@@ -1530,6 +1532,14 @@ def _activation_needs_beta(cls, activation):
 
     @classmethod
     def _impl_v7(cls, inputs, attr, params):
+        if cls.name == 'LSTM':
```
Comment (areusch) [inline python/tvm/relay/frontend/onnx.py:1535]: here I think it'd be cleaner to just implement the `@classmethod` in each subclass, no?

File: python/tvm/relay/frontend/onnx.py [hunk 2]
```diff
@@ -1530,6 +1532,14 @@ def _activation_needs_beta(cls, activation):
...
+            return cls._lstm(inputs, attr, params)
+        if cls.name == 'GRU':
+            return cls._gru(inputs, attr, params)
+        raise NotImplementedError("%s RNNs are not yet supported." % cls.name)
+
+    @classmethod
+    def _lstm(cls, inputs, attr, params):
```
Comment (areusch) [inline python/tvm/relay/frontend/onnx.py:1542]: any reason not to move this to LSTM subclass?

File: python/tvm/relay/frontend/onnx.py (hunk 1 above)
Comment (jwfromm) [reply python/tvm/relay/frontend/onnx.py]: you're right thats a nicer organization. I've updated the structure accordingly.

**Comment by jwfromm** [comment]
I've added a small change to how arguments are parsed in RNNs based on the discussion here: https://discuss.tvm.ai/t/onnx-lstm-op-conversion/7238/5. Our previous implementation assumed that the position of optional arguments could not be known without their name, however, this is not true assuming the nodes are constructing properly. To avoid name based problems, RNNs now use indexing to get inputs. I've updated the `onnx_input` structure to allow indexing outside of the input bounds but return `None` in such cases, indicating that an input was not provided.

**Comment by masahi** [comment]
Thanks @jwfromm @areusch @anijain2305

**Pull request merged**